### PR TITLE
[MAINTENANCE] Only entities set up in Configuration must be needed to instantiate primary Profiler typed objects

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Changing the signatures of `ParameterBuilder` and `Rule` classes so as to accommodate instantiation from Configuration.
* Ability to look up parameters for a single `Domain` scope and cross-`Domain` scope.
* Refactor of `get_parameter_value()` for elegance and code reuse.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

File: great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py [hunk 1]
```diff
@@ -1,15 +1,13 @@
...
 import great_expectations.exceptions as ge_exceptions
 from great_expectations import DataContext
 from great_expectations.core.batch import BatchRequest
-from great_expectations.profiler.domain_builder.domain import Domain
-from great_expectations.profiler.domain_builder.util import get_batch_ids
 from great_expectations.profiler.parameter_builder.parameter_builder import (
     ParameterBuilder,
 )
-from great_expectations.validator.validator import Validator
+from great_expectations.profiler.util import get_batch_ids
```
Comment (anthonyburdi) [inline great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py:10]: Is this import needed?

**Review by anthonyburdi — APPROVED** [review]
LGTM - no need to re-review if that import is not needed

File: great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py:10]: @anthonyburdi This import is used on line 39.

Comment (anthonyburdi) [reply great_expectations/profiler/parameter_builder/multi_batch_parameter_builder.py:10]: My mistake! I did not expand the un-edited code.

File: great_expectations/profiler/util.py [hunk 2]
```diff
@@ -25,66 +75,99 @@ def get_parameter_value(
...
+            if parameter_value is not None:
+                return parameter_value
+        except KeyError as e:
+            exception_messages.append(
+                f"""Unable to find value for parameter name "{fully_qualified_parameter_name}" for domain_id \
+"{domain_id}": "{e}" occurred.
+"""
+            )
+    if len(exception_messages) > 0:
+        raise ge_exceptions.ProfilerExecutionError(message=";".join(exception_messages))
```
Comment (anthonyburdi) [inline great_expectations/profiler/util.py]: I love this collecting of messages for a more comprehensive error!

File: great_expectations/profiler/util.py [hunk 3]
```diff
@@ -25,66 +75,99 @@ def get_parameter_value(
...
+    for domain_id in domain_ids:
+        parameter_container = parameters[domain_id]
+        try:
+            parameter_value = _get_parameter_value_one_domain_scope(
+                fully_qualified_parameter_name=fully_qualified_parameter_name,
+                fully_qualified_parameter_name_as_list=fully_qualified_parameter_name_as_list,
+                parameters=parameter_container,
+            )
+            # In the present implementation, the first non-NULL parameter value found is returned.  In the future,
+            # parameter name across domains will need to be disambiguated (e.g., bu using a domain-specific scoping).
```
Comment (anthonyburdi) [inline great_expectations/profiler/util.py]: Thank you for adding this note for our future selves.

File: great_expectations/profiler/util.py [hunk 4]
```diff
@@ -14,61 +17,157 @@
 )
 
 
+def domain_kwargs_parameter(func: Callable = None) -> Callable:
+    @wraps(func)
```
Comment (talagluck) [inline great_expectations/profiler/util.py]: I would love to better understand the reasoning behind this decorator

Comment (alexsherstinsky) [reply great_expectations/profiler/util.py]: It has been removed.

**Review by talagluck — APPROVED** [review]
LGTM!

**Review by anthonyburdi — APPROVED** [review]
LGTM!

File: great_expectations/profiler/util.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/profiler/util.py]: 🙇

**Pull request merged**